data/transaction/status.go

```diff
@@ -0,0 +1,83 @@
+package transaction
+
+import (
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/data/block"
+)
+
+// TxStatus is the status of a transaction
+type TxStatus string
+
+const (
+	// TxStatusReceived = received but not yet executed
+	TxStatusReceived TxStatus = "received"
+	// TxStatusPartiallyExecuted = received and executed on source shard
+	TxStatusPartiallyExecuted TxStatus = "partially-executed"
+	// TxStatusExecuted = received and executed
+	TxStatusExecuted TxStatus = "executed"
+	// TxStatusNotExecuted = received and executed with error
+	// Question for review: not executed means "executed with error", correct?
+	TxStatusNotExecuted TxStatus = "not-executed"
+	// TxStatusInvalid = considered invalid
+	TxStatusInvalid TxStatus = "invalid"
+)
+
+// StatusComputer computes a transaction status
+type StatusComputer struct {
+	MiniblockType    block.Type
+	SourceShard      uint32
+	DestinationShard uint32
+	Receiver         []byte
+	TransactionData  []byte
+	SelfShard        uint32
+}
+
+// ComputeStatusWhenInPool computes the transaction status when transaction is in pool
+func (params *StatusComputer) ComputeStatusWhenInPool() TxStatus {
+	if params.isContractDeploy() {
```

Comment:
why is this check needed ?

Reply:
`params.isDestinationMe() && params.isCrossShard()` would be evaluated to `true` for deployments, but still we don't want to return `TxStatusPartiallyExecuted` in that case (seen in previous PR with the status for deployments).

Does it make sense?

Reply:
isCrossShard should return false in case of deployment.

Reply:
Currently, `isCrossShard()` is implemented as follows:

```
func (params *StatusComputer) isCrossShard() bool {
	return params.SourceShard != params.DestinationShard
}
```

While `ComputeStatusWhenInPool()` is fed the parameters as follows - in a generic manner, not knowing anything special about the Zero address:

```
SourceShard:      n.shardCoordinator.ComputeId(tx.Tx.GetSndAddr()),
DestinationShard: n.shardCoordinator.ComputeId(tx.Tx.GetRcvAddr()),
Receiver:         tx.Tx.GetRcvAddr(),
TransactionData:  tx.Data,
SelfShard:        n.shardCoordinator.SelfId(),
```

Thus, the `isContractDeploy()` has to be checked first (because the check `isDestinationMe() && isCrossShard()` will be fooled otherwise).

We could change `isCrossShard()` to call `isContractDeploy()`, but perhaps that would be a bit harder to reason about in the future.

Fixed to be more explicit, as before. Is it all right?

Reply:
in this case shardCoordinator computeID should have returned selfID. for the empty address.

Reply:
in computestatuswheninPool you might use another function a wrapper on top of computeID where you add the empty check.

Reply:
Let's keep compute ID completely out of the `statusComputer`, please.

Previous location, previous PR (reviewed):
https://github.com/ElrondNetwork/elrond-go/pull/2234/files

Reply:
Sometimes it happens that these things get approved by a PR and the next one requests changes in the same zone. It would be better to have correct shardId computation at the base. In this case shard coordinator, or computestatus

Reply:
Left as it is, as discussed separately.